Add thread-local-state NamesMode and NoNamesGuard

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24474 Implement name inference for torch.dot
* #24471 Implement name inference for addmv, addmv_, mv
* #24469 Implement name inference for expand
* #24306 Implement name inference for mm, addmm, addmm_
* #24921 Fix binary op name inference between unnamed and named tensors.
* #24914 Add helper function Tensor::names()
* #24907 Rename Tensor::names() to Tensor::opt_names()
* #24885 Fix `FIXME_default_names` by storing static list of 64 none names
* **#24367 Add thread-local-state NamesMode and NoNamesGuard**
* #24203 Implement name inference for t(), transpose(...)

NamesMode determines whether or not to ignore the names field of
TensorImpl. In particular, when it is disabled, all tensors are treated
as unnamed.

Test Plan:
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 24367 gh/zou3519/105/head

Differential Revision: [D16915806](https://our.internmc.facebook.com/intern/diff/D16915806)

zou3519 committed:
Add thread-local-state NamesMode and NoNamesGuard

NamesMode determines whether or not to ignore the names field of
TensorImpl. In particular, when it is disabled, all tensors are treated
as unnamed.

Test Plan:
- New tests [namedtensor ci]

zou3519 committed:
Update on "Add thread-local-state NamesMode and NoNamesGuard"

Add thread-local-state NamesMode and NoNamesGuard

NamesMode determines whether or not to ignore the names field of
TensorImpl. In particular, when it is disabled, all tensors are treated
as unnamed.

Test Plan:
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 24367 gh/zou3519/105/head

nairbv commented on aten/src/ATen/NamedTensor.h:
```diff
@@ -47,6 +47,26 @@ struct CAFFE2_API NamedTensorMeta : public c10::NamedTensorMetaInterface {
+};
+
+// A RAII, thread local (!) guard that enables or disables names upon
+// construction, and sets it back to the original value upon destruction.
```
might be helpful to also describe why this is needed

zou3519 replied:
That's good to have in general. I'm not sure what the guidelines for "how to best use NoNamesGuard" is right now, so I didn't write that in.

zou3519 committed:
Update on "Add thread-local-state NamesMode and NoNamesGuard"

Add thread-local-state NamesMode and NoNamesGuard

NamesMode determines whether or not to ignore the names field of
TensorImpl. In particular, when it is disabled, all tensors are treated
as unnamed.

Test Plan:
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 24367 gh/zou3519/105/head

zou3519 committed:
Update on "Add thread-local-state NamesMode and NoNamesGuard"

Add thread-local-state NamesMode and NoNamesGuard

NamesMode determines whether or not to ignore the names field of
TensorImpl. In particular, when it is disabled, all tensors are treated
as unnamed.

Test Plan:
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 24367 gh/zou3519/105/head
